events/events_network.c

```diff
@@ -292,30 +323,47 @@ struct eventrec *
 events_network_get(void)
 {
 	struct eventrec * r;
-	size_t nfds = socketlist_getsize(S);
 
 	/* We haven't found any events yet. */
 	r = NULL;
 
+	/* We have no fds, thus we have no events. */
+	if (fds == NULL)
+		return (r);
+
 	/* Scan through the fd sets looking for ready sockets. */
 	for (; fdscanpos < nfds; fdscanpos++) {
+		struct pollfd * current = &fds[fdscanpos];
+
+		/* File descriptor was closed, ignore it in future polls. */
+		if (current->revents & POLLNVAL) {
+			current->fd = -1;
+			continue;
+		}
+
 		/* Are we ready for reading? */
-		if (FD_ISSET(fdscanpos, &readfds)) {
-			r = socketlist_get(S, fdscanpos)->reader;
-			socketlist_get(S, fdscanpos)->reader = NULL;
+		if (current->revents & POLLIN) {
+			r = socketlist_get(S, (size_t)current->fd)->reader;
```

Comment:
... actually, it's even worse than that: Imagine the sequence:
1. _select gets called
2. _cancel gets called
3. the socket is closed
4. a new socket is opened
5. _register gets called
6. _get is called

In that case we'd be reporting an event *on a completely different socket* to the one which was previously detected as ready.  We need some sort of validity marker which _cancel can clear, or possibly a freshness value we can compare between the event registration and when poll was called.